db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -522,21 +549,42 @@ public static Integer isIndexUseful(Predicate pred,
             return null;
 
         /* Look for an index column on one side of the relop */
-        if (baseColumnPositions != null) {
-            for (indexPosition = 0; indexPosition < baseColumnPositions.length; indexPosition++) {
-                if (isIn) {
-                    if (inNode.getLeftOperand() instanceof ColumnReference) {
-                        indexCol = (ColumnReference) inNode.getLeftOperand();
-                        if ((optTable.getTableNumber() != indexCol.getTableNumber())
-                                || (indexCol.getColumnNumber() != baseColumnPositions[indexPosition])
-                                || inNode.selfReference(indexCol))
-                            indexCol = null;
+        if (isIndexOnExpr) {
+            LanguageConnectionContext lcc = pred.getLanguageConnectionContext();
+            CompilerContext newCC = lcc.pushCompilerContext();
+            Parser p = newCC.getParser();
+
+            String[] exprTexts = indexDescriptor.getExprTexts();
+            for (indexPosition = 0; indexPosition < exprTexts.length; indexPosition++) {
```

Comment:
Yes, we definitely need to cache them. I have been thinking about it and created a Jira ticket to track things that could be improved: https://splicemachine.atlassian.net/browse/DB-10312.

I didn't do it in this PR because I was not sure about where to cache these ASTs. I was thinking that they would be required in many other places not touched in this change, and if I implement something back then, it probably needs to be changed later.

Now that I'm almost in the last major piece of this whole story, my idea is to cache the index expression ASTs in `ConglomerateDescriptor`. They will not be serialized or deserialized with other fields. Instead, we have always an empty `Valuenode` array when getting an instance either by constructing or deserialzing. Index expression will be parsed and stored in the array in the first time we use them, and later we just get from the array.

I think it's a good place because we set `ConglomerateDescriptor` in access path once and almost every following operations on this access path get this instance. It's global enough, yet much better scoped than the context.

I'd like to have your opinions on this before I start the implementation.

Reply:
Thanks @ascend1 ! It makes sense to me to cache it in the ConglomerateDescriptor. I couldn't think of any side effect of this approach, but it would be great to run through @dgomezferro and @jyuanca also and see if they see any downside about this approach.